### PR TITLE
fix: hide Approve/Reject on already-handled follow-request notifications

### PR DESCRIPTION
**Pull request by llun**
## Problem

On the Notifications page, a **follow request** row kept showing **Approve / Reject** actions even after the request was already handled — e.g. the viewer had approved it and the requester now follows them (see the reported screenshot, `@null@llun.dev` still showing Approve/Reject). Clicking the actions on a non-pending request just results in a `404` from the follow-request API, so the buttons were misleading.

### Root cause

`FollowRequestNotification` always initialised its local state to `'pending'`, so it rendered the action buttons unconditionally, independent of the real `Follow` record state on the server.

## Fix

The notifications Server Component now resolves the live follow state for each `follow_request` row and forwards it as `followRequestStatus`:

| Live `Follow` status | Resolved state | Row renders |
| --- | --- | --- |
| `Requested` | `pending` | Approve / Reject actions |
| `Accepted` | `accepted` | "Approved" label, no actions |
| Rejected / Undo / no record | `resolved` | "No longer pending" label, no actions |

- The follower is the notification's **source actor**; the viewer is the **target**. The lookup reuses the existing, well-tested `getAcceptedOrRequestedFollow`.
- `FollowRequestNotification` seeds its initial state from `initialStatus` (defaulting to `'pending'` so a row without a resolved status still offers the actions), and keeps the in-session `'rejected'` state produced by its own Reject action.
- The status→UI mapping is extracted into a pure `followRequestStatusFromFollow` helper. The `default` branch maps any non-`Requested`/`Accepted` follow to `resolved`, so a stale/unexpected status can never resurface the buttons.

## Tests

- `followRequestStatus.test.ts` — pure mapper: `null`/`Requested`/`Accepted`/`Rejected`/`Undo` → expected UI state.
- `FollowRequestNotification.test.tsx` — renders actions for `pending`, "Approved" + no actions for `accepted`, "No longer pending" + no actions for `resolved`.
- `NotificationItem.test.tsx` — threading: an `accepted` / `resolved` row renders the label and no Approve/Reject buttons.

## Verification

- `yarn run prettier --write .` ✓
- `yarn lint` ✓ (0 errors)
- `yarn build` ✓
- `yarn test` ✓ (570 files, 5048 tests)

No migrations changed, so the reference schema dumps are untouched.

### CODE REVIEW

File: app/(timeline)/notifications/NotificationItem.tsx [hunk 1]
```diff
@@ -128,7 +132,12 @@ export const NotificationItem = ({
       if (notification.type === 'follow') {
         body = <FollowNotification account={acc} />
       } else if (notification.type === 'follow_request') {
-        body = <FollowRequestNotification account={acc} />
+        body = (
+          <FollowRequestNotification
+            account={acc}
+            initialStatus={notification.followRequestStatus}
+          />
+        )
```
[the remark this answers is not in the document]
Comment (llun) [reply app/(timeline)/notifications/NotificationItem.tsx:146]: Addressed in 3f72403c, but keyed by the **resolved status** rather than a stable entity id — keying by `acc.url` / `notification.id` would not actually reset the state here:

- `NotificationItem` is already keyed by `notification.id` in `NotificationsList`, so its sole `body` child (`FollowRequestNotification`) is never remounted across a soft `router.refresh()` while the id stays constant. A constant-per-row key (id/url) therefore leaves a row that was handled elsewhere still showing stale Approve/Reject after the mark-as-read `router.refresh()`. The new regression test `NotificationItem › re-seeds the follow-request row when the resolved status changes on refresh` fails with an id/url key and passes with the status key.
- Keying by `followRequestStatus` does **not** destroy in-flight operations: this prop only changes when the **server** re-renders, which happens *after* the async accept/reject resolves (`router.refresh()` is called on success), and the failure path never refreshes. So a pending request or the inline error is never interrupted by a key change — the key changes only once the server state has actually changed, which is exactly when a re-seed is wanted.

[the remark this answers is not in the document]
Comment (llun) [reply app/(timeline)/notifications/NotificationItem.tsx:146]: Addressed in 3f72403c, but keyed by the **resolved status** rather than a stable entity id — keying by `acc.url` / `notification.id` would not actually reset the state here:

- `NotificationItem` is already keyed by `notification.id` in `NotificationsList`, so its sole `body` child (`FollowRequestNotification`) is never remounted across a soft `router.refresh()` while the id stays constant. A constant-per-row key (id/url) therefore leaves a row that was handled elsewhere still showing stale Approve/Reject after the mark-as-read `router.refresh()`. The new regression test `NotificationItem › re-seeds the follow-request row when the resolved status changes on refresh` fails with an id/url key and passes with the status key.
- Keying by `followRequestStatus` does **not** destroy in-flight operations: this prop only changes when the **server** re-renders, which happens *after* the async accept/reject resolves (`router.refresh()` is called on success), and the failure path never refreshes. So a pending request or the inline error is never interrupted by a key change — the key changes only once the server state has actually changed, which is exactly when a re-seed is wanted.

**Comment by llun** [comment]
### Sub-agent review — round 1

Ran a multi-dimensional adversarial review (correctness, edge cases, types, tests, conventions, a11y/UX), each finding independently verified. **2 confirmed (both addressed), 2 dismissed.**

**Confirmed & addressed:**
1. **Resolve by `notification.followId`, not the requester/viewer pair** (correctness). A follow_request notification records the exact `Follow` it concerns (`createFollower.ts` sets `followId`). Resolving by the `(sourceActorId, targetActorId)` pair mis-handles a re-request after a rejection (the old rejected row would leak its actions onto the new request). Now resolves via `getFollowFromId(followId)` (falling back to the pair lookup only when no `followId` is recorded), and surfaces a faithful `rejected` state. → commit c8988e92
2. **Untested live accept/reject + page wiring** (tests). Added interaction tests (`Approve→Approved`, `Reject→Rejected`, failure path) so `STATUS_LABEL.rejected` and the optimistic action are load-bearing, plus `page.test.tsx` asserting the page resolves follow state by `followId` only for `follow_request` rows. → commit c8988e92

**Dismissed (with reason):**
- *“`getAcceptedOrRequestedFollow` error fails the whole page”* — pre-existing behavior; every other per-row DB read in this `Promise.all` (account/status/collection) already propagates, so the page-level throw is the established pattern, not a regression introduced here.
- *“`followRequestStatus.test.ts` Rejected/Undo cases are defensive-only”* — these exercise the mapper’s `default`/`rejected` branches and are worth keeping; not a defect.

Gemini’s `key` suggestion was also addressed (keyed by resolved status, not id — see resolved threads). → commit 3f72403c

**Comment by llun** [comment]
### Sub-agent review — round 2 (on the followId resolver + key-by-status)

Re-ran the adversarial review across resolver/key/correctness/tests/conventions. **1 confirmed (addressed), 1 dismissed.**

**Confirmed & addressed:**
- **`page.test.tsx` pinned that the resolver is *called* but not that its result is *forwarded*.** Removing the `followRequestStatus` forward from the page (compute-then-discard, the exact user-visible failure) kept both page tests green. Hardened the test to walk the returned element tree and assert the follow_request row reaches `NotificationsList` with `followRequestStatus: 'accepted'` (and non-follow_request rows carry none). → commit 30999174

**Dismissed (with reason):**
- *"Fallback path (no followId → `getAcceptedOrRequestedFollow`) not exercised through `Page`"* — already fully covered by the `resolveFollowRequestStatus` unit test, which pins the exact `{actorId: sourceActorId, targetActorId: viewer}` argument shape; the page→resolver call is a single direct pass-through. Re-pinning it at the page boundary would only duplicate working coverage.

No unresolved review threads; Kilo: "No Issues Found / Merge".

File: app/(timeline)/notifications/followRequestStatus.ts [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+  viewerActorId: string
+): Promise<FollowRequestInitialStatus> => {
+  const follow = notification.followId
+    ? await database.getFollowFromId({ followId: notification.followId })
+    : await database.getAcceptedOrRequestedFollow({
+        actorId: notification.sourceActorId,
+        targetActorId: viewerActorId
+      })
+  return followRequestStatusFromFollow(follow)
+}
```
[the remark this answers is not in the document]
Comment (llun) [reply app/(timeline)/notifications/followRequestStatus.ts:65]: Adopted in bc766f94 — the lookups in `resolveFollowRequestStatus` are now wrapped in `try/catch`, logging via the server `logger` — with one change: the fallback is `'resolved'`, not `'pending'`. Falling back to `'pending'` would re-show Approve/Reject for a request that may already be accepted (re-introducing exactly the stale-action bug this PR fixes), and clicking those would 404. `'resolved'` hides the actions safely; the row self-heals on the next load. Covered by a new test asserting the resolver returns `'resolved'` and logs when the lookup throws.

**Comment by llun** [comment]
### Sub-agent review — round 3

Re-ran the adversarial review (correctness, resilience/logging, tests, conventions/types) against the latest code including the `try/catch` resilience change. **0 findings** — no dimension surfaced an actionable issue.

Requesting a confirming bot round.

/gemini review

File: app/(timeline)/notifications/components/FollowRequestNotification.tsx [hunk 3]
```diff
@@ -3,24 +3,42 @@
...
+  initialStatus = 'pending'
+}) => {
   const router = useRouter()
-  const [status, setStatus] = useState<'pending' | 'accepted' | 'rejected'>(
-    'pending'
-  )
+  const [status, setStatus] =
+    useState<FollowRequestInitialStatus>(initialStatus)
   const [isLoading, setIsLoading] = useState(false)
   const [error, setError] = useState<string | null>(null)
```
[the remark this answers is not in the document]
Comment (llun) [reply app/(timeline)/notifications/components/FollowRequestNotification.tsx:43]: Keeping the single `isLoading` boolean here — the concurrent-execution concern is already covered and the state machine would be unused granularity:

- **Concurrent actions are already prevented:** both Approve and Reject are `disabled={isLoading}` for the duration of a request, and React flushes the disabled state between discrete click events, so a second/cross action can't start. A same-tick double-fire of the *same* action is harmless anyway — once the follow is no longer `Requested`, the authorize/reject route returns 404, and `setStatus` is idempotent.
- **Matches the established pattern:** the sibling two-action row `CollectionConsentNotification` uses the identical `isLoading` boolean with both buttons `disabled={isLoading}`. Introducing an `'accepting' | 'rejecting' | null` machine only here would diverge from it.
- **YAGNI:** a per-action status would only earn its keep if the row rendered a per-button loading indicator, which it doesn't — so the extra state would be unused.

Resolving as a reasoned no-change.

[the remark this answers is not in the document]
Comment (llun) [reply app/(timeline)/notifications/components/FollowRequestNotification.tsx:43]: Adopted in 7d7b6c77 — added a synchronous `useRef` re-entrancy lock: `respond()` early-returns if `pendingRef.current` is set, sets it before starting, and clears it in `finally`. This closes the same-tick gap that `disabled={isLoading}` alone can't (the `isLoading` closure is stale within a tick), while `isLoading` stays for the visual/disabled state. Added a test asserting a second action does not start while one is in flight.

**Comment by llun** [comment]
### Sub-agent review — round 4 (on the `useRef` re-entrancy lock)

Re-ran the adversarial review (re-entrancy, correctness, tests/conventions). **1 confirmed (addressed), 0 dismissed.**

**Confirmed & addressed:**
- **The in-flight test wasn't load-bearing.** It `await`ed the `isLoading` re-render before the follow-up clicks, so `disabled={isLoading}` (the pre-existing guard) — not the new `pendingRef` lock — swallowed them; the test stayed green even with the lock removed. Rewrote it to reproduce the same-tick race: both clicks dispatched in one `act()` batch *before* React flushes the disabled state, so only the synchronous `pendingRef` lock can stop the second call. **Verified it fails when the lock is removed.** Split the disabled-buttons check into its own focused test. → commit 4b899d5e

The re-entrancy lock itself was found correct. Gemini's latest round returned no feedback; no unresolved threads.

**Comment by llun** [comment]
### Sub-agent review — final holistic pass

Ran a final merge-blocker + coverage review across the whole change. **0 merge-blockers; 1 low coverage gap (addressed).**

**Confirmed & addressed:**
- **Retry-after-failure was untested**, so the `respond()` `finally` reset (releasing `isLoading` + the `pendingRef` lock) wasn't load-bearing — a regression moving those resets into the success branch would permanently lock the buttons after any failure yet keep every test green. Added a test that fails a request, asserts the action re-enables, then retries successfully and asserts the second dispatch swaps to the approved label. **Verified it fails when the resets are moved out of `finally`.** → commit 7d117854

Everything else (mapper incl. Rejected, by-id vs pair fallback, lookup-failure→resolved, page `followId` wiring + forwarding, all four UI states, same-tick re-entrancy, re-seed-on-status-change) was confirmed well-covered and load-bearing. No merge-blockers.
